### PR TITLE
Drop SHARD_STATE_TO_DELETE

### PR DESCRIPTION
**Pull request by agedemenli**
DESCRIPTION: Drop `SHARD_STATE_TO_DELETE` and use the cleanup records instead

Drops the shard state that is used to mark shards as orphaned. Now we insert cleanup records into `pg_dist_cleanup` so "orphaned" shards will be dropped either by maintenance daemon or internal cleanup calls. With this PR, we make the "cleanup orphaned shards" functions to be no-op, as they would not be needed anymore.

This PR includes some naming changes about placement functions. We don't need functions that filter orphaned shards, as there will be no orphaned shards anymore.

We will also be introducing a small script with this PR, for users with orphaned shards. We'll basically delete the orphaned shard entries from `pg_dist_placement` and insert cleanup records into `pg_dist_cleanup` for each one of them, during Citus upgrade.

We also have a lot of flakiness fixes in this PR.

### CODE REVIEW

File: src/backend/distributed/sql/citus--11.1-1--11.2-1.sql [hunk 1]
```diff
@@ -13,3 +13,12 @@ DROP FUNCTION pg_catalog.worker_append_table_to_shard(text, text, text, integer)
 #include "udfs/citus_task_wait/11.2-1.sql"
 #include "udfs/citus_prepare_pg_upgrade/11.2-1.sql"
 #include "udfs/citus_finish_pg_upgrade/11.2-1.sql"
+
+-- drop orphaned shards after inserting records for them into pg_dist_cleanup
```
Comment (agedemenli) [inline src/backend/distributed/sql/citus--11.1-1--11.2-1.sql:17]: Should we do anything for downgrades?
What if the user downgrades to a version that `pg_dist_cleanup` doesn't exist?

Comment (JelteF) [reply src/backend/distributed/sql/citus--11.1-1--11.2-1.sql:17]: I think the best is to fail the downgrade

File: src/backend/distributed/operations/shard_cleaner.c [hunk 2]
```diff
@@ -112,57 +110,25 @@ static int CompareCleanupRecordsByObjectType(const void *leftElement,
...
 
 /*
- * isolation_cleanup_orphaned_shards implements a test UDF that's the same as
- * citus_cleanup_orphaned_shards. The only difference is that this command can
- * be run in transactions, this is to test
+ * isolation_cleanup_orphaned_shards is noop.
+ * Use isolation_cleanup_orphaned_resources instead.
  */
 Datum
 isolation_cleanup_orphaned_shards(PG_FUNCTION_ARGS)
```
Comment (JelteF) [inline src/backend/distributed/operations/shard_cleaner.c:130]: This one we can remove. We never included the SQL definition in a release. It was only created during tests.

File: src/backend/distributed/operations/shard_cleaner.c [hunk 3]
```diff
@@ -234,8 +214,7 @@ TryDropOrphanedResources(bool waitForLocks)
 
 	PG_TRY();
 	{
-		droppedResourceCount = DropOrphanedShardsForMove(waitForLocks);
-		droppedResourceCount += DropOrphanedResourcesForCleanup();
+		droppedResourceCount = DropOrphanedResourcesForCleanup();
```
Comment (JelteF) [inline src/backend/distributed/operations/shard_cleaner.c:217]: I didn't realise that we don't have waitForLocks for the new function. Any clue why the old one had this parameter? Should we have it too for the new function?

File: src/backend/distributed/sql/citus--11.1-1--11.2-1.sql [hunk 4]
```diff
@@ -13,3 +13,12 @@ DROP FUNCTION pg_catalog.worker_append_table_to_shard(text, text, text, integer)
...
+INSERT INTO pg_dist_cleanup
+    SELECT 0, 0, 0, shard_name(sh.logicalrelid, sh.shardid) AS object_name, plc.groupid AS node_group_id, 0
+        FROM pg_dist_placement plc
+        JOIN pg_dist_shard sh ON sh.shardid = plc.shardid
+        WHERE plc.shardstate = 4;
+
+DELETE FROM pg_dist_placement WHERE shardstate = 4;
```
Comment (JelteF) [inline src/backend/distributed/sql/citus--11.1-1--11.2-1.sql:24]: Can we add a test for this?

File: src/backend/distributed/operations/shard_transfer.c [hunk 5]
```diff
@@ -413,6 +416,45 @@ citus_move_shard_placement(PG_FUNCTION_ARGS)
...
+	ListCell *shardIntervalCell = NULL;
+
+	foreach(shardIntervalCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		ListCell *shardPlacementCell = NULL;
+		uint64 oldShardId = shardInterval->shardId;
+
+		/* mark for deferred drop */
+		List *shardPlacementList = ActiveShardPlacementList(oldShardId);
```
Comment (JelteF) [inline src/backend/distributed/operations/shard_transfer.c:435]: I think this function should take the group id of the node from which shard is moved. Currently if replication_factor is more than 1 for the shard, it means that a shard move would remove both active placements.

File: src/backend/distributed/operations/shard_cleaner.c (hunk 3 above)
Comment (agedemenli) [reply src/backend/distributed/operations/shard_cleaner.c:217]: >  Before doing any of this it will take an exclusive PlacementCleanup lock.
 This is to ensure that this function is not being run concurrently.
 Otherwise really bad race conditions are possible, such as removing all
 placements of a shard. waitForLocks indicates if this function should
 wait for this lock or not.

Seems like it was to prevent dropping the placements that shouldn't be dropped. But with the `pg_dist_cleanup` records, we have the node group and object name so it doesn't look like it would cause such an issue. Thoughts?

File: src/backend/distributed/operations/shard_transfer.c (hunk 5 above)
Comment (agedemenli) [reply src/backend/distributed/operations/shard_transfer.c:437]: just introduced a new function that does the same thing but only for the placements on a certain node

File: src/backend/distributed/operations/shard_transfer.c [hunk 6]
```diff
@@ -413,6 +418,82 @@ citus_move_shard_placement(PG_FUNCTION_ARGS)
...
+	ShardInterval *shardInterval = NULL;
+	foreach_ptr(shardInterval, shardIntervalList)
+	{
+		/* mark for deferred drop */
+		List *shardPlacementList =
+			ActiveShardPlacementListOnGroup(shardInterval->shardId, groupId);
+
+		ShardPlacement *placement = NULL;
+		foreach_ptr(placement, shardPlacementList)
+		{
```
Comment (JelteF) [inline src/backend/distributed/operations/shard_transfer.c:478]: No need for these lines (and the second loop) anymore, since we already have the groupId and the shardInterval

File: src/test/regress/bin/normalize.sed [hunk 7]
```diff
@@ -307,3 +307,9 @@ s/(NOTICE:  issuing SET LOCAL application_name TO 'citus_rebalancer gpid=)[0-9]+
 
 # PG13 changes bgworker sigterm message, we can drop that line with PG13 drop
 s/(FATAL: terminating).*Citus Background Task Queue Executor.*(due to administrator command)\+/\1 connection \2                    \+/g
+
+# shard_rebalancer output, flaky numbers for total_bytes
+s/public( *)\|( *)tab(.*)\|(.*)\|( *)[0-9][0-9][0-9][0-9][0-9][0-9][0-9]/public\1\|\2tab\3\|\4\|\5xxxxxxx/g
+s/public( *)\|( *)tab(.*)\|(.*)\|( *)[0-9][0-9][0-9][0-9][0-9][0-9]/public\1\|\2tab\3\|\4\|\5xxxxxx/g
```
Comment (JelteF) [inline src/test/regress/bin/normalize.sed:313]:
```suggestion
s/public( *)\|( *)tab(.*)\|(.*)\|( *)[0-9]+/public\1\|\2tab\3\|\4\|\5xxxxxxx/g
```

File: src/test/regress/bin/normalize.sed [hunk 8]
```diff
@@ -307,3 +307,9 @@ s/(NOTICE:  issuing SET LOCAL application_name TO 'citus_rebalancer gpid=)[0-9]+
...
+# shard_rebalancer output, flaky improvement number
+s/improvement of 0.1[0-9][0-9][0-9][0-9][0-9] is lower/improvement of 0.1xxxxx is lower/g
```
Comment (JelteF) [inline src/test/regress/bin/normalize.sed:315]:
```suggestion
s/improvement of 0.1[0-9]* is lower/improvement of 0.1xxxxx is lower/g
```

File: src/test/regress/citus_tests/run_test.py [hunk 9]
```diff
@@ -47,6 +50,9 @@
         )
         sys.exit(1)
 
+if test_file_name in test_files_to_skip:
```
Comment (JelteF) [inline src/test/regress/citus_tests/run_test.py:53]: Let's print a warning, like is don for arbitrary config tests in line 84

File: src/test/regress/sql/cpu_priority.sql [hunk 10]
```diff
@@ -71,7 +71,9 @@ SET citus.cpu_priority_for_logical_replication_senders = 15;
 SELECT master_move_shard_placement(11568900, 'localhost', :worker_2_port, 'localhost', :worker_1_port, 'force_logical');
 SET citus.max_high_priority_background_processes = 3;
 SELECT master_move_shard_placement(11568900, 'localhost', :worker_1_port, 'localhost', :worker_2_port, 'force_logical');
-CALL citus_cleanup_orphaned_shards();
+SET client_min_messages TO ERROR;
```
Comment (JelteF) [inline src/test/regress/sql/cpu_priority.sql:74]: Why are these necessary?  I thought wait_for_resource_cleanup already removed logs. i.e. what errors is this showing?

(same for these same statements in other tests files)

Comment (agedemenli) [reply src/test/regress/sql/cpu_priority.sql:74]: not errors but warnings like "cleanup failed for 1 object out of 4"

File: src/backend/distributed/operations/shard_cleaner.c (hunk 3 above)
Comment (marcocitus) [reply src/backend/distributed/operations/shard_cleaner.c:208]: We now (always) lock at the operation ID level, where we previously had a global lock.

One high-level confusing thing is that if 2 processes run cleanup at the same time, they will race for operation IDs. As in, they might both get some arbitrary subset of operation IDs, but they will be distinct subsets.

File: src/backend/distributed/operations/shard_transfer.c [hunk 11]
```diff
@@ -376,14 +376,33 @@ citus_move_shard_placement(PG_FUNCTION_ARGS)
 		PlacementMovedUsingLogicalReplicationInTX = true;
 	}
 
+	DropOrphanedResourcesInSeparateTransaction();
```
Comment (marcocitus) [inline src/backend/distributed/operations/shard_transfer.c:379]: This seems like a behavioural change. I'm kind of wondering whether we really want this (e.g. what if this fails on some unrelated node?), and whether we want it here (e.g. if we have parallel moves, should this be in rebalancer_start?).

Comment (agedemenli) [reply src/backend/distributed/operations/shard_transfer.c:379]: Does it have a downside to have it here? We take locks on operationIds during the cleanup anyway. Besides, the commands we send to the workers to drop the objects are optional. What could be the problem in case it fails to drop some object?

File: src/backend/distributed/operations/shard_cleaner.c [hunk 12]
```diff
@@ -195,10 +152,26 @@ citus_cleanup_orphaned_resources(PG_FUNCTION_ARGS)
 }
 
 
+/*
+ * isolation_cleanup_orphaned_resources implements a test UDF that's the same as
+ * citus_cleanup_orphaned_resources. The only difference is that this command can
+ * be run in transactions, this is to test
```
Comment (JelteF) [inline src/backend/distributed/operations/shard_cleaner.c:158]: missing part of the comment (it already had this problem in the original comment that you copy pasted).
```suggestion
 * be run in transactions, this is needed to test this function in isolation tests since commands are automatically run in transactions there.
```

File: src/test/regress/spec/isolation_move_placement_vs_move_placement.spec [hunk 13]
```diff
@@ -46,6 +56,12 @@ step "s2-commit"
 	COMMIT;
 }
 
+step "s2-wait-and-commit"
+{
+	SELECT pg_sleep(1);
```
Comment (JelteF) [inline src/test/regress/spec/isolation_move_placement_vs_move_placement.spec:61]: Why is this needed?

Comment (agedemenli) [reply src/test/regress/spec/isolation_move_placement_vs_move_placement.spec:61]: otherwise session2 immediately commits and releases the lock on the shard, so the shard move gets completed with the cleanup

File: src/backend/distributed/operations/shard_transfer.c (hunk 11 above)
Comment (JelteF) [reply src/backend/distributed/operations/shard_transfer.c:380]: I think this is indeed fine to do since the commands that are run are optional. It seems desirable, to allow for back and forth movement of shards.

File: src/backend/distributed/operations/shard_transfer.c [hunk 14]
```diff
@@ -397,7 +416,7 @@ citus_move_shard_placement(PG_FUNCTION_ARGS)
 	}
 
 	/* since this is move operation, we remove shards from source node after copy */
```
Comment (JelteF) [inline src/backend/distributed/operations/shard_transfer.c:418]: comment seems outdated
```suggestion
	/* since this is move operation, we remove the placements from the metadata for the source node after copy */
```

File: src/backend/distributed/operations/shard_transfer.c (hunk 11 above)
Comment (JelteF) [reply src/backend/distributed/operations/shard_transfer.c:380]: Also, we already did this sometimes in `EnsureShardCanBeCopied` (and it's removed there now).

File: src/backend/distributed/operations/shard_transfer.c [hunk 15]
```diff
@@ -1839,27 +1897,22 @@ RecreateTableDDLCommandList(Oid relationId)
...
- * MarkForDropColocatedShardPlacement marks the shard placement metadata for
- * the given shard placement to be deleted in pg_dist_placement. The function
- * does this for all colocated placements.
+ * DropColocatedShardPlacements drops the shard placement metadata for
+ * the shard placements of given colocated shard interval list from pg_dist_placement.
  */
 static void
-MarkForDropColocatedShardPlacement(ShardInterval *shardInterval, char *nodeName,
-								   int32 nodePort)
+DropColocatedShardPlacements(List *colocatedShardList,
```
Comment (JelteF) [inline src/backend/distributed/operations/shard_transfer.c:1904]: nit: This function doesn't call ColocatedShardIntervalList anymore. So there's no reason why it should have colocated in its name or arguments (or comment). It simple deletes all placement rows for the given nodes for the given shard list, no matter if they are colocated or not. Also maybe to make clearer that this only deletes the placement from the metadata and not from the actual node, we could include something like `FromMetadata` in the function name.
```suggestion
DropShardPlacementsFromMetadata(List *shardList,
```

File: src/test/regress/bin/normalize.sed [hunk 16]
```diff
@@ -297,3 +297,11 @@ s/"Total Cost": [0-9].[0-9]+/"Total Cost": xxxx/g
 
 # normalize gpids
 s/(NOTICE:  issuing SET LOCAL application_name TO 'citus_rebalancer gpid=)[0-9]+/\1xxxxx/g
+
+# PG13 changes bgworker sigterm message, we can drop that line with PG13 drop
+s/(FATAL: terminating).*Citus Background Task Queue Executor.*(due to administrator command)\+/\1 connection \2                    \+/g
```
Comment (JelteF) [inline src/test/regress/bin/normalize.sed:305]: This line seems to be part of line 120 of this same file. Was this some incorrect merge conflict resolution?

File: src/test/regress/sql/cpu_priority.sql (hunk 10 above)
Comment (JelteF) [reply src/test/regress/sql/cpu_priority.sql:74]: Hmm, okay. Any reason wait_for_resource_cleanup  shows warnings? Sounds like it would be better to hide those also.

File: src/test/regress/bin/normalize.sed (hunk 16 above)
Comment (agedemenli) [reply src/test/regress/bin/normalize.sed]: hmm, yes, fixed now. Thanks

File: src/test/regress/bin/normalize.sed [hunk 17]
```diff
@@ -37,6 +37,9 @@ s/ daily_uniques_[0-9]+ / daily_uniques_xxxxxxx /g
 # shard table names for isolation_create_citus_local_table
 s/"citus_local_table_([0-9]+)_[0-9]+"/"citus_local_table_\1_xxxxxxx"/g
 
+# shard table names for isolation_move_placement_vs_move_placement
+s/ public.test_move_table_[0-9]+/public.test_move_table_xxxxxxx/g
```
Comment (JelteF) [inline src/test/regress/bin/normalize.sed:41]: Instead of this it's better to make the shard number fixed by using in the `setup` phase of the isolation test.
```sql
ALTER SEQUENCE pg_catalog.pg_dist_shardid_seq RESTART <random number>;
```

File: src/test/regress/expected/isolation_rebalancer_deferred_drop.out [hunk 18]
```diff
@@ -40,12 +38,11 @@ master_move_shard_placement
...
 step s2-drop-marked-shards:
     SET client_min_messages to DEBUG1;
-    CALL isolation_cleanup_orphaned_shards();
+    CALL isolation_cleanup_orphaned_resources();
 
-s1: NOTICE:  cleaned up 1 orphaned shards
 step s1-drop-marked-shards:
-    SET client_min_messages to NOTICE;
-    CALL isolation_cleanup_orphaned_shards();
+    SET client_min_messages to ERROR;
```
Comment (JelteF) [inline src/test/regress/expected/isolation_rebalancer_deferred_drop.out:44]: Why was this needed? It seems that the NOTICE output was kind of intended in these tests.

File: src/test/regress/expected/multi_colocated_shard_rebalance.out [hunk 19]
```diff
@@ -910,3 +993,14 @@ DROP TABLE table1_group1;
...
+DROP SCHEMA move_partitions CASCADE;
+-- this is because issue:
+-- https://github.com/citusdata/citus/issues/6580
+-- we can remove this once the issue is fixed
+SELECT 1 FROM run_command_on_workers($$DROP TABLE IF EXISTS test_with_pkey_13000043$$);
+ ?column?
+---------------------------------------------------------------------
+        1
+        1
+(2 rows)
```
Comment (JelteF) [inline src/test/regress/expected/multi_colocated_shard_rebalance.out:1005]: It seems like this can be dropped now?

File: src/test/regress/expected/multi_extension.out [hunk 20]
```diff
@@ -1198,6 +1235,30 @@ SELECT * FROM multi_extension.print_extension_changes();
...
+(2 rows)
+
+-- error out as cleanup records remain
+ALTER EXTENSION citus UPDATE TO '11.0-4';
+ERROR:  pg_dist_cleanup is introduced in Citus 11.1
+HINT:  To downgrade Citus to an older version, you should first cleanup all the orphaned resources and make sure pg_dist_cleanup is empty, by executing CALL citus_cleanup_orphaned_resources();
+CONTEXT:  PL/pgSQL function inline_code_block line XX at RAISE
+-- cleanup
+DELETE FROM pg_dist_cleanup;
+DELETE FROM pg_dist_shard WHERE shardid % 2 = 1;
```
Comment (JelteF) [inline src/test/regress/expected/multi_extension.out:1260]: Afaict this will leave the orphaned shard tables on the node. It seems better to run what the error message suggests:
```
CALL citus_cleanup_orphaned_resources()
```

File: src/test/regress/expected/shard_rebalancer.out [hunk 21]
```diff
@@ -1151,18 +1149,18 @@ SELECT * FROM public.table_placements_per_node;
...
-         146 |  123023 |          1 |           0 |      14
-         149 |  123024 |          1 |           0 |      14
-         152 |  123027 |          1 |           0 |      14
-         153 |  123028 |          1 |           0 |      14
-         154 |  123021 |          1 |           0 |      16
-         155 |  123025 |          1 |           0 |      16
-         156 |  123022 |          1 |           0 |      16
-         157 |  123026 |          1 |           0 |      16
+CALL citus_cleanup_orphaned_resources();
+select shardid, shardstate, shardlength from pg_dist_placement ORDER BY placementid;
```
Comment (JelteF) [inline src/test/regress/expected/shard_rebalancer.out:1153]: Why filter out groupid and placementid?

File: src/test/regress/expected/shard_rebalancer.out [hunk 22]
```diff
@@ -1324,12 +1322,12 @@ SELECT * from master_set_node_property('localhost', :worker_2_port, 'shouldhaves
 
 (1 row)
 
-SELECT * FROM get_rebalance_table_shards_plan(threshold := 0, drain_only := true);
+SELECT * FROM get_rebalance_table_shards_plan(threshold := 0, drain_only := true) ORDER BY shardid;
```
Comment (JelteF) [inline src/test/regress/expected/shard_rebalancer.out:1325]: Why was this ORDER BY needed? normally get_rebalance_table_shards_plan should return a consistent order.

File: src/test/regress/expected/shard_rebalancer.out [hunk 23]
```diff
@@ -1585,8 +1582,8 @@ WHERE table_schema = 'public'
 ) a ORDER BY table_name;
  table_schema | table_name | row_estimate | total_bytes
 ---------------------------------------------------------------------
- public       | tab_123041 |        10000 |      368640
- public       | tab_123043 |        10000 |      368640
+ public       | tab_123041 |        10000 |      xxxxxxx
+ public       | tab_123043 |        10000 |      xxxxxxx
```
Comment (JelteF) [inline src/test/regress/expected/shard_rebalancer.out:1586]: These sizes are important for the intent of these tests. I'm guessing you changed this to avoid test flakyness? If so there's another way of making this not flaky, that will keep the intent of the test. It's documented here: https://github.com/citusdata/citus/blob/main/src/test/regress/flaky_tests.md#disk-size-numbers-are-not-exactly-like-expected

File: src/test/regress/sql/multi_test_helpers.sql [hunk 24]
```diff
@@ -169,7 +169,7 @@ $$ LANGUAGE plpgsql;
 
 -- Introduce a function that waits until all cleanup records are deleted, for testing purposes
 CREATE OR REPLACE FUNCTION wait_for_resource_cleanup() RETURNS void
-SET client_min_messages TO WARNING
+SET client_min_messages TO ERROR
```
Comment (JelteF) [inline src/test/regress/sql/multi_test_helpers.sql:172]: Now that you added this, you should be able to remove `SET client_min_messages TO ERROR` in lots of places where you added it in this PR.

File: src/test/regress/sql/shard_rebalancer.sql [hunk 25]
```diff
@@ -1427,3 +1428,17 @@ RESET client_min_messages;
...
+\c - - - :worker_1_port
+SET citus.enable_ddl_propagation TO OFF;
+REVOKE ALL ON SCHEMA public FROM testrole;
+DROP USER testrole;
+
+\c - - - :worker_2_port
+SET citus.enable_ddl_propagation TO OFF;
+REVOKE ALL ON SCHEMA public FROM testrole;
+DROP USER testrole;
+DROP TABLE test_rebalance_with_disabled_worker_433500, test_rebalance_with_disabled_worker_433501, test_rebalance_with_disabled_worker_433502, test_rebalance_with_disabled_worker_433503;
```
Comment (JelteF) [inline src/test/regress/sql/shard_rebalancer.sql:1444]: Is it expected that these were not removed?

File: src/test/regress/bin/normalize.sed [hunk 26]
```diff
@@ -297,3 +297,8 @@ s/"Total Cost": [0-9].[0-9]+/"Total Cost": xxxx/g
 
 # normalize gpids
 s/(NOTICE:  issuing SET LOCAL application_name TO 'citus_rebalancer gpid=)[0-9]+/\1xxxxx/g
+
+# shard_rebalancer output, flaky numbers for total_bytes
+s/public( *)\|( *)tab(.*)\|(.*)\|( *)[0-9]+/public\1\|\2tab\3\|\4\|\5xxxxxxx/g
```
Comment (JelteF) [inline src/test/regress/bin/normalize.sed:302]: Like said in another comment, lets remove this rule by using a better size query.

File: src/test/regress/expected/isolation_rebalancer_deferred_drop.out (hunk 18 above)
Comment (JelteF) [reply src/test/regress/expected/isolation_rebalancer_deferred_drop.out:44]: decision after discussion: let's add a `select * from pg_dist_cleanup where type = 'placement` call after this. Because the warning messages are not consistent anymore now that this also cleans up replication slots/subscriptions.

File: src/test/regress/expected/shard_rebalancer.out [hunk 27]
```diff
@@ -1595,7 +1592,7 @@ SELECT * FROM get_rebalance_table_shards_plan('tab');
 ---------------------------------------------------------------------
 (0 rows)
 
-SELECT * FROM get_rebalance_table_shards_plan('tab', rebalance_strategy := 'by_disk_size');
+SELECT * FROM get_rebalance_table_shards_plan('tab', rebalance_strategy := 'by_disk_size') ORDER BY shardid;
```
Comment (JelteF) [inline src/test/regress/expected/shard_rebalancer.out:1595]: This one is not necessary for sure.

File: src/test/regress/expected/shard_rebalancer.out (hunk 22 above)
Comment (JelteF) [reply src/test/regress/expected/shard_rebalancer.out:1325]: Add comment that we're, sorting this plan because it involves different colocation groups and which colocation group is first moved is not consistent.

File: src/test/regress/sql/shard_rebalancer.sql (hunk 25 above)
Comment (agedemenli) [reply src/test/regress/sql/shard_rebalancer.sql:1484]: yes, they are left on a disabled node. We didn't notice it until this pr. Now we get errors because of them, in the flaky check

File: src/test/regress/spec/isolation_move_placement_vs_move_placement.spec [hunk 28]
```diff
@@ -46,6 +67,11 @@ step "s2-commit"
 	COMMIT;
 }
 
+step "s2-wait-and-commit"
+{
+	COMMIT;
+}
```
Comment (JelteF) [inline src/test/regress/spec/isolation_move_placement_vs_move_placement.spec:73]: As discussed over chat this is now the same as the `s2-commit` step right above. So let's remove this one.

**Review by JelteF — APPROVED** [review]

File: src/test/regress/sql/shard_rebalancer.sql (hunk 25 above)
Comment (JelteF) [reply src/test/regress/sql/shard_rebalancer.sql:1484]: The main thing I'm wondering is if this is intended or if this is a bug. No need to fix it in this PR, but if it's a bug we should at least open an issue.

File: src/test/regress/sql/shard_rebalancer.sql [hunk 29]
```diff
@@ -902,17 +930,23 @@ WHERE table_schema = 'public'
 ) a ORDER BY table_name;
 
 \c - - - :master_port
-SELECT * FROM get_rebalance_table_shards_plan('tab', rebalance_strategy := 'by_disk_size');
+SELECT * FROM get_rebalance_table_shards_plan('tab', rebalance_strategy := 'by_disk_size') ORDER BY shardid;
```
Comment (JelteF) [inline src/test/regress/sql/shard_rebalancer.sql:933]: Needs a comment on why this needs order by. I guess in this case it's because the disk sizes of the shard groups are normally the same, but sometimes randomly one is slightly bigger than the other, resulting in a different shard group being moved first.

File: src/test/regress/sql/shard_rebalancer.sql [hunk 30]
```diff
@@ -706,69 +707,72 @@ SELECT * from master_set_node_property('localhost', :worker_2_port, 'shouldhaves
...
 SELECT * FROM rebalance_table_shards(threshold := 0, shard_transfer_mode := 'block_writes');
-CALL citus_cleanup_orphaned_shards();
+CALL citus_cleanup_orphaned_resources();
 SELECT * FROM public.table_placements_per_node;
 
 -- testing behaviour when setting shouldhaveshards to false and rebalancing all
 -- colocation groups with drain_only=false
 SELECT * from master_set_node_property('localhost', :worker_2_port, 'shouldhaveshards', false);
-SELECT * FROM get_rebalance_table_shards_plan(threshold := 0);
+SELECT * FROM get_rebalance_table_shards_plan(threshold := 0) ORDER BY shardid;
```
Comment (JelteF) [inline src/test/regress/sql/shard_rebalancer.sql:747]:
```suggestion
-- we actually shouldn't need the ORDER BY clause as the output will be in execution order
-- but this one involves different colocation groups and which colocation group is first moved is not consistent
SELECT * FROM get_rebalance_table_shards_plan(threshold := 0) ORDER BY shardid;
```
